Acdp 224 (#7)
Actindo-AG/Documentation: ActindoWorkFlow/Operation/01_ManageWorkflows.md

[!!Workflows](ActindoWorkFlow)

# Manage the workflows

The *Workflows* module allows the customers to define their own business processes, such as orders, shipments, returns, etc.
The workflows can be created individually in the workflow editor, they can be edited subsequently and by publishing and unpublishing a workflow, the usage can be controlled. Further, obsolete workflows can be deleted.

## Create a workflow

Create a workflow to define a new process.

### Prerequisites

No prerequisites to fulfill.

### Procedure

*Workflows > Workflows > Tab OVERVIEW*

![Workflows](/Assets/Screenshots/ActindoWorkFlow/Workflows/Workflows.png "[Workflows]")

1. Click the ![Add](/Assets/Icons/Plus01.png "[Add]") (Add) button in the bottom right corner.   
  The *New Workflow* window is displayed.

  ![New workflow](/Assets/Screenshots/ActindoWorkFlow/Workflows/NewWorkflow.png "[New workflow]")

2. Enter a name for the new workflow in the *Select a name for your new workflow* field.

> 3. Enter a key for the workflow in the *Select a unique key for your new workflow* field. The key is required for API access and must be system wide unique.

Reviewer: ```suggestion
3. Enter a key for the workflow in the *Select a unique key for your new workflow* field. The key is required for API access and must be unique within the workflow version.
```

  A new place connected with an arc to the selected output port is displayed.

  ![First place](/Assets/Screenshots/ActindoWorkFlow/Workflows/FirstPlace.png "[First place]")

2. Add further actions by following the procedure [Create an action](#create-an-action) or connect the place with the end place as described in the following step.

  > [Info] A workflow must contain at least one action and can contain an unlimited number of actions.

3. Click the place, drag it over the place to which you want to connect it and drop it.     
  The places are merged and the new connection is created.

### Next steps

- [Edit a workflow](#edit-a-workflow)
- [Publish a workflow](#publish-a-workflow)
- [Unpublish a workflow](#unpublish-a-workflow)
- [Delete a workflow](#delete-a-workflow)

### See also

[Workflow and process elements](/ActindoWorkFlow/Overview/02_WorkflowProcessElements.md)

## Edit a workflow

Edit a workflow to modify elements within the workflow, to extend the workflow by adding further actions or to reduce the workflow by removing actions.

### Prerequisites

At least one workflow is created, see [Create a workflow](#create-a-workflow).

> 

Author: add prerequisite:
rights to edit a workflow

> At least one workflow is created, see [Create a workflow](#create-a-workflow).

Author: ```suggestion
- At least one workflow is created, see [Create a workflow](#create-a-workflow).
- You have the required rights to edit a workflow.
```

### Procedure

*Workflows > Workflows > Tab OVERVIEW*

![Workflows](/Assets/Screenshots/ActindoWorkFlow/Workflows/Workflows.png "[Workflows]")

1. Click the workflow you want to edit in the list of workflows.   
  The *Workflow Versions* view with all versions of the selected workflow is displayed.

  ![Workflow versions](/Assets/Screenshots/ActindoWorkFlow/Workflows/WorkflowVersions.png "[Workflow versions]")

> 2. Click the workflow version you want to edit in the list of workflow versions.  

Author: add corresponding comment for the future

[!!Workflows](ActindoWorkFlow)

# Manage the workflows

The *Workflows* module allows the customers to define their own business processes, such as orders, shipments, returns, etc.

> The workflows can be created individually in the workflow editor, they can be edited subsequently and by publishing and unpublishing a workflow, the usage can be controlled. Further, obsolete workflows can be deleted.

Reviewer: ```suggestion
The workflows can be created individually in the workflow editor, they can be edited subsequently and by publishing and unpublishing a workflow, the usage can be controlled. Further, obsolete workflows can be deleted.
```

Stimmt der letzte Satz? Ich habe gedacht in dieser Version kann man noch nichts löschen.

  > [Info] The list of places is filtered for your keyword as you type.

5. Click the start place you want to add in the list of places.    

6. Click the *Choose the data type of your end place* field and enter the name of the place or a keyword you are searching for.

  > [Info] The list of places is filtered for your keyword as you type.

7. Click the end place you want to add in the list of places.  

8. Click the [CREATE] button in the bottom right corner.   
  The new workflow is created. The *New Workflow* window is closed. The workflow editor with the defined start and end places is displayed.  

  ![Workflow editor new](/Assets/Screenshots/ActindoWorkFlow/Workflows/WorkflowEditorNew.png "[Workflow editor new]")

#### Create a transition

1. Click the ![Add](/Assets/Icons/Plus04.png "[Add]") (Add) button next to the place or click the [NEW ACTION] button on the right hand above the workflow editor.     
  A window to search for a transition is displayed.

  ![Search action](/Assets/Screenshots/ActindoWorkFlow/Workflows/SearchAction.png "[Search action]")

> 2. Click the *search for an action* field and enter the name of the transition or a keyword you are searching for. If you have clicked the ![Add](/Assets/Icons/Plus04.png "[Add]") (Add) button next to a place, only those transitions that are compatible with the data type of the place are displayed in the list.

Reviewer: In Ui ist es klein, aber ich habe es generell großgeschrieben (Verweise). Einheitlich machen?

3. Click the place, drag it over the place to which you want to connect it and drop it.     
  The places are merged and the new connection is created.

### Next steps

- [Edit a workflow](#edit-a-workflow)
- [Copy a workflow](#copy-a-workflow)
- [Publish a workflow version](#publish-a-workflow-version)
- [Unpublish a workflow version](#unpublish-a-workflow-version)
- [Track a workflow process](02_TrackProcesses.md)

### See also

- [User Interface workflows](/ActindoWorkFlow/UserInterface/00_UserInterface.md)
- [Workflow and process elements](/ActindoWorkFlow/Overview/02_WorkflowProcessElements.md)

## Edit a workflow

Edit a workflow to modify elements within the workflow, to extend the workflow by adding further transitions or to reduce the workflow by removing transitions.

### Prerequisites

- At least one workflow is created, see [Create a workflow](#create-a-workflow).
- You have the required rights to edit a workflow.

### Procedure

*Workflows > Workflows > Tab OVERVIEW*

![Workflows](/Assets/Screenshots/ActindoWorkFlow/Workflows/Workflows.png "[Workflows]")

1. Click the workflow you want to edit in the list of workflows.   

>   The *Workflow Versions* view with all versions of the selected workflow is displayed.

Reviewer: ```suggestion
  The *Workflow Versions* view with all versions of the selected workflow is displayed.
```The *Workflow versions*...

Wieder das Thema mit groß/klein. Ob es sich lohnt bzw. ob es zeitlich zu schaffen ist, für diesen Release das zu vereinheitlichen...

  ![Add action](/Assets/Screenshots/ActindoWorkFlow/Workflows/AddAction.png "[Add action]")

#### Edit a transition

1. Click the transition you want to edit.  
  The transition is highlighted and its settings are displayed in the settings side bar on the right.

  ![Action](/Assets/Screenshots/ActindoWorkFlow/Workflows/Action.png "[Action]")

  > [Info] The settings displayed in the settings side bar may differ depending on the respective transition. The sections *Configuration* and *Static Inputs* are only available for certain transitions.     

2. Edit the settings field in the settings side bar. You can change the label, the queue type, the task event, the configuration and the static inputs as described below:

  + Click the *Label* field and edit the name of the transition displayed in the workflow editor.

  + Click the *Queue Type* drop-down list and select the appropriate queue type for the transition. All available queue types are displayed in the drop-down list. By default the *Default* queue type is selected. You can define a different queue type for each transition within the workflow.   

>   For detailed information about the queue types, see [Configure the queue types](/ActindoWorkFlow/Integration/01_WConfigureQueueTypes.md).

Reviewer: ```suggestion
  For detailed information about the queue types, see [Configure the queue types](/ActindoWorkFlow/Integration/01_ConfigureQueueTypes.md).
```

>   + Click the *Task Event* drop-down list and select an event connected to the transition. The event is triggered and the corresponding task is assigned when the process action is failed. All available task events are displayed in the drop-down list. By default, the *ActindoWorkflow Process Execution Failed* event is preselected. The drop-down list is only displayed when the *Task* module is installed and the current user has the required rights to configure it.

Reviewer: ```suggestion
  + Click the *Task Event* drop-down list and select an event connected to the transition. The event is triggered and the corresponding task is assigned when the process action has failed. All available task events are displayed in the drop-down list. By default, the *ActindoWorkflow Process Execution Failed* event is preselected. The drop-down list is only displayed when the *Task* module is installed and the current user has the required rights to configure it.
```

  + Edit the field(s) in the *Configuration* section to define further configuration settings to the selected transition. This section is only displayed for certain transitions. The fields displayed in the *Configuration* section depend on the selected transition.

  + Click the ![Add](/Assets/Icons/Plus04.png "[Add]") (Add) button right to an input name in the *Static Inputs* section to add a static data value to the corresponding input. All input ports that are not yet connected to a place are displayed in this section. Enter the appropriate value in the text field below the input name. The value entered must be a valid JSON value. Click the ![Delete](/Assets/Icons/Trash07.png "[Delete]") (Delete) button right to a value to delete it. This section is only displayed for certain transitions.

>   > [Info] Input ports with an static input value cannot be connected to a place. Therefore, a small arrow without a place is displayed in front of an input port with a static input value. Delete the static input value from the port to be able to connect it with a place.     

Reviewer: ```suggestion
  > [Info] Input ports with a static input value cannot be connected to a place. Therefore, a small arrow without a place is displayed in front of an input port with a static input value. Delete the static input value from the port to be able to connect it with a place.     
```

3. Click the ![Points](/Assets/Icons/Points02.png "[Points]") (Points) button in the upper left corner next to the workflow name.   
  The workflow context menu is displayed.

  ![Context menu](/Assets/Screenshots/ActindoWorkFlow/Workflows/ContextMenu.png "[Context menu]")

4. Click the *Export JSON ..* menu entry in the context menu.   
  The *Export JSON* window with the JSON code for the current workflow is displayed.

  ![Export JSON](/Assets/Screenshots/ActindoWorkFlow/Workflows/ExportJSON.png "[Export JSON]")

5. Select the complete JSON code in the window, copy it to the clipboard and click the [CLOSE] button in the bottom right corner.   
  The *Export JSON* window is closed.

6. Click the ![Back](/Assets/Icons/Back02.png "[Back]") (Back) button in the workflow editor.  
  The workflow editor is closed and the list of workflows is displayed again.

#### Import the workflow JSON code

*Workflows > Workflows > Tab OVERVIEW*

![Workflows](/Assets/Screenshots/ActindoWorkFlow/Workflows/Workflows.png "[Workflows]")

1. [Create a new workflow](#create-a-workflow) or click the workflow in the list of workflows to which you want to import the copied JSON code.   

>   The workflow editor is displayed.

Reviewer: ``The workflow editor is displayed.

Wenn man auf einem Workflow klickt, kommt (in dieser Version noch) Workflow versions, oder? Dann auf Version klicken und dann hat man den Editor.

  ![Workflow editor new](/Assets/Screenshots/ActindoWorkFlow/Workflows/WorkflowEditorNew.png "[Workflow editor new]")

> [comment]: <> (Step 2 as well as the screenshot workflow versions is possibly not needed in the next workflows version as the version view will disappear and clicking the workflow will take you to the editor directly, check next time)

Reviewer: ``Step 2 as well as the screenshot workflow versions is possibly not needed in the next workflows version as the version view will disappear and clicking the workflow will take you to the editor directly, check next time

Aktuell für dieses Release. Wenn nicht, Aufzählung aktualisieren (3 -> 2, usw.)

  The *Import JSON* window is closed. The copied workflow is displayed in the workflow editor.

  ![Workflow editor copied](/Assets/Screenshots/ActindoWorkFlow/Workflows/WorkflowEditorCopied.png "[Workflow editor copied]")

  > [Info] You can edit the workflow as desired before deploying it.

6. Click the ![Points](/Assets/Icons/Points02.png "[Points]") (Points) button in the upper left corner next to the workflow name.   
  The workflow context menu is displayed.

  ![Context menu](/Assets/Screenshots/ActindoWorkFlow/Workflows/ContextMenu.png "[Context menu]")

7. Click the *Deploy* menu entry in the context menu.   
  The copied workflow is deployed and published.

### Next steps

- [Publish a workflow version](#publish-a-workflow-version)
- [Unpublish a workflow version](#unpublish-a-workflow-version)
- [Track a workflow process](02_TrackProcesses.md)

### See also

- [User Interface workflows](/ActindoWorkFlow/UserInterface/00_UserInterface.md)
- [Create a workflow](#create-a-workflow)
- [Edit a workflow](#edit-a-workflow)
- [Workflow and process elements](/ActindoWorkFlow/Overview/02_WorkflowProcessElements.md)

## Publish a workflow version

> A workflow can only be used for processes if it is published. If multiple versions of a workflow exist, the highest published version is taken for upcoming workflow processes. Therefore you have to publish a workflow to use it for workflow processes. By default, a created workflow or a deployed workflow version is automatically published. You can manually publish or [unpublish](#unpublish-a-workflow-version) workflow versions.

Reviewer: `By default, a **created workflow or a deployed workflow** version is automatically published`

Das klingt für mich ein wenig verwirrend, vielleicht das "or". Vielleicht "and"? Was ist damit genau gemeint? Ist ein "created workflow" automatisch veröffentlicht, also die erste Version des Workflows? Denn man kann "versions" (und nicht workflows) veröffentlichen, oder?

### Prerequisites

- At least one workflow is created, see [Create a workflow](#create-a-workflow).
- At least one workflow version is unpublished, see [Unpublish a workflow version](#unpublish-a-workflow-version).

### Procedure

*Workflows > Workflows > Tab OVERVIEW > Select workflow*

![Workflow versions published](/Assets/Screenshots/ActindoWorkFlow/Workflows/WorkflowVersionsPublished.png "[Workflow versions published]")

1. Select the checkbox of the workflow version you want to publish.      
  The editing toolbar is displayed.

> 2. Click the [UNPUBLISH] button in the editing toolbar.   

Reviewer: ``Click the [UNPUBLISH] button in the editing toolbar. 

Sollte es nicht [PUBLISH] sein? Abschnitt heißt "Publish a workflow version". Ähnliches gilt für nächsten Abschnitt. Anderum oder habe ich falsch verstanden?

  The selected workflow version is unpublished. The icon in the *Published* column changes from ![Checkmark](/Assets/Icons/Check.png "[Checkmark]") (Checkmark) to ![Cross](/Assets/Icons/Cross02.png "[Cross]") (Cross).

  ![Workflow versions unpublished](/Assets/Screenshots/ActindoWorkFlow/Workflows/WorkflowVersionsUnpublished.png "[Workflow versions unpublished]")

### Next steps

- [Unpublish a workflow version](#unpublish-a-workflow-version)
- [Track a workflow process](02_TrackProcesses.md)

### See also

- [User Interface workflows](/ActindoWorkFlow/UserInterface/00_UserInterface.md)
- [Create a workflow](#create-a-workflow)
- [Edit a workflow](#edit-a-workflow)
- [Copy a workflow](#copy-a-workflow)
- [Workflow and process elements](/ActindoWorkFlow/Overview/02_WorkflowProcessElements.md)

## Unpublish a workflow version

> A workflow can only be used for processes if it is published. By default, a created workflow or a deployed workflow version is automatically published. You can manually [publish](#publish-a-workflow-version) or unpublish workflow versions. Unpublish a workflow version if the workflow is still in work or you use testing purposes it and you the workflow should not yet be used for processes.

Reviewer: ```suggestion
A workflow can only be used for processes if it is published. By default, a created workflow or a deployed workflow version is automatically published. You can manually [publish](#publish-a-workflow-version) or unpublish workflow versions. Unpublish a workflow version if the workflow is still in work or you use it for testing purposes and the workflow should not yet be used for processes.
```

### Prerequisites

- At least one workflow is created, see [Create a workflow](#create-a-workflow).
- At least one workflow version is published, see [Publish a workflow version](#publish-a-workflow-version).

### Procedure

*Workflows > Workflows > Tab OVERVIEW*

  ![Workflow versions unpublished](/Assets/Screenshots/ActindoWorkFlow/Workflows/WorkflowVersionsUnpublished.png "[Workflow versions unpublished]")

1. Select the checkbox of the workflow version you want to unpublish.      
  The editing toolbar is displayed.

2. Click the [PUBLISH] button in the editing toolbar.   

>   The selected workflow version is unpublished. The icon in the *Published* column changes from ![Cross](/Assets/Icons/Cross02.png "[Cross]") (Cross) to ![Checkmark](/Assets/Icons/Check.png "[Checkmark]") (Checkmark).

Reviewer: ```suggestion
  The selected workflow version is published. The icon in the *Published* column changes from ![Cross](/Assets/Icons/Cross02.png "[Cross]") (Cross) to ![Checkmark](/Assets/Icons/Check.png "[Checkmark]") (Checkmark).
```
